File: app/src/main/java/com/giftinapp/business/MerchantActivity.java
package com.giftinapp.business;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentSender;
import android.os.Bundle;
import android.util.Log;
import android.util.SparseArray;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.giftinapp.business.business.GiftACustomerFragment;
import com.giftinapp.business.business.GiftinAboutForMerchant;
import com.giftinapp.business.business.MerchantGiftStatsFragment;
import com.giftinapp.business.business.MerchantInfoUpdate;
import com.giftinapp.business.business.RateInfluencerFragment;
import com.giftinapp.business.business.SetRewardDeal;
import com.giftinapp.business.business.WalletInfo;
import com.giftinapp.business.customer.BrandPreferenceFragment;
import com.giftinapp.business.customer.MerchantStoryList;
import com.giftinapp.business.utility.SessionManager;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.play.core.appupdate.AppUpdateInfo;
import com.google.android.play.core.appupdate.AppUpdateManager;
import com.google.android.play.core.appupdate.AppUpdateManagerFactory;
import com.google.android.play.core.install.model.AppUpdateType;
import com.google.android.play.core.install.model.InstallStatus;
import com.google.android.play.core.install.model.UpdateAvailability;
import com.google.firebase.FirebaseApp;
import com.google.firebase.appcheck.FirebaseAppCheck;
import com.google.firebase.appcheck.safetynet.SafetyNetAppCheckProviderFactory;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreSettings;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.squareup.picasso.Picasso;
import com.synnapps.carouselview.CarouselView;
import com.synnapps.carouselview.ViewListener;

import java.util.List;
import java.util.Objects;

import co.paystack.android.PaystackSdk;

public class MerchantActivity extends AppCompatActivity {
    BottomNavigationView bottomNavigation;

    public SessionManager sessionManager;

    private FirebaseAuth mAuth = FirebaseAuth.getInstance();

    protected CarouselView carouselViewMerchant;

    protected SparseArray<MerchantReportsViewHolder> holderListMerchant = new SparseArray<>();

    public Integer numberOfCustomerGifted = null;
    public Long totalAmountGifted = null;
    public Long totalWalletBalance = null;

    private DrawerLayout drawer;
    private ActionBarDrawerToggle t;
    private NavigationView nv;

    AppUpdateManager appUpdateManager;

    public Integer counter = 0;

    public Integer following = 0;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_merchant);


        appUpdateManager = AppUpdateManagerFactory.create(this);

        PaystackSdk.initialize(getApplicationContext());
        // Returns an intent object that you use to check for an update.
        com.google.android.play.core.tasks.Task<AppUpdateInfo> appUpdateInfoTask = appUpdateManager.getAppUpdateInfo();

        // Checks that the platform will allow the specified type of update.
        appUpdateInfoTask.addOnSuccessListener(appUpdateInfo -> {
            if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE
                    // This example applies an immediate update. To apply a flexible update
                    // instead, pass in AppUpdateType.FLEXIBLE
                    && appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE)) {
                // Request the update.

                try {
                    appUpdateManager.startUpdateFlowForResult(
                            // Pass the intent that is returned by 'getAppUpdateInfo()'.
                            appUpdateInfo,
                            // Or 'AppUpdateType.FLEXIBLE' for flexible updates.
                            AppUpdateType.IMMEDIATE,
                            // The current activity making the update request.
                            this,
                            // Include a request code to later monitor this update request.
                            1);
                } catch (IntentSender.SendIntentException e) {
                    e.printStackTrace();
                }

            }
        });

        sessionManager = new SessionManager(getApplicationContext());

        getNumberOfFollowers();

        drawer = findViewById(R.id.merchantNavDrawerLayout);
        t = new ActionBarDrawerToggle(this, drawer,R.string.navigation_drawer_open, R.string.navigation_drawer_close);

        drawer.addDrawerListener(t);
        t.syncState();

        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        nv = findViewById(R.id.merchantNavView);

        nv.setNavigationItemSelectedListener(item -> {
            selectDrawerItem(item);
            return true;
        });

        View headerView = nv.getHeaderView(0);
        TextView navTextView = headerView.findViewById(R.id.nav_header_textView);
        ImageView navImageView = headerView.findViewById(R.id.nav_header_imageView);
//        Picasso.get().load(R.drawable.ic_brandible_icon).into(navImageView);
        navTextView.setText(Objects.requireNonNull(mAuth.getCurrentUser()).getEmail());

        carouselViewMerchant = findViewById(R.id.carouselView);

        carouselViewMerchant.setPageCount(3);
        carouselViewMerchant.setViewListener(viewListener);

        carouselViewMerchant.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                updateCounter(position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
    }

    ViewListener viewListener = position -> {
        @SuppressLint("InflateParams") View customView = getLayoutInflater().inflate(R.layout.single_item_merchant_carousel_report,null);

        MerchantReportsViewHolder holder = new MerchantReportsViewHolder();
        holder.reportValue = customView.findViewById(R.id.kpi_report_value);
        holder.reportName = customView.findViewById(R.id.kpi_report_name);
        holder.reportIcon = customView.findViewById(R.id.kpi_report_icon);




        switch (position) {

            case 0: {
                getNumberOfCustomersGifted();
                holder.reportName.setText("Total Influencers Rewarded");
                holder.reportIcon.setImageResource(R.drawable.happycustomer);
                long totalGiftCoinSum= numberOfCustomerGifted==null ? 0 : numberOfCustomerGifted;
                holder.reportValue.setText(String.valueOf(totalGiftCoinSum));
                holderListMerchant.put(0, holder);
                break;
            }

//            case 1: {
//                getTotalAmountGifted();
//                holder.reportName.setText("Total Amount Gifted");
//                holder.reportValue.setText(String.valueOf(totalAmountGifted));
//                holder.reportIcon.setImageResource(R.drawable.ic_gifts);
//                holderListMerchant.put(1, holder);
//                break;
//            }

            case 1: {
                getWalletBalance();
                holder.reportName.setText("Wallet Balance");
                holder.reportValue.setText(String.valueOf(totalWalletBalance));
                holder.reportIcon.setImageResource(R.drawable.gift_coin_icon);
                holderListMerchant.put(1, holder);
                break;
            }

            case 2: {
                getNumberOfFollowers();
                holder.reportName.setText("Influencers following your Brand");
                holder.reportValue.setText(String.valueOf(sessionManager.getFollowingCount()));
                holder.reportIcon.setImageResource(R.drawable.influencer_following_brand_icon);
                holderListMerchant.put(2, holder);
                break;
            }

        }

        return customView;
    };

    private void updateCounter(int position) {

        switch (position) {
            case 0: {
                getNumberOfCustomersGifted();
                long totalGiftCoinSum= numberOfCustomerGifted==null ? 0L : numberOfCustomerGifted;
                holderListMerchant.get(0).reportValue.setText(String.valueOf(totalGiftCoinSum));
                break;
            }
//            case 1: {
//                getTotalAmountGifted();
//                long totalAmountGiftedCustomer= totalAmountGifted==null ? 0L : totalAmountGifted;
//                holderListMerchant.get(1).reportValue.setText(String.valueOf(totalAmountGiftedCustomer));
//                break;
//            }

            case 1: {
                getWalletBalance();
                long walletBalanceTotal= totalWalletBalance==null ? 0L : totalWalletBalance;
                holderListMerchant.get(1).reportValue.setText(String.valueOf(walletBalanceTotal));
                break;
            }

            case 2: {
                getNumberOfFollowers();
                int numberOfFollowers= sessionManager.getFollowingCount();
                holderListMerchant.get(2).reportValue.setText(String.valueOf(numberOfFollowers));
                break;
            }
        }
    }

    public static class MerchantReportsViewHolder {
        TextView reportValue;
        TextView reportName;
        ImageView reportIcon;
    }


    private void getWalletBalance() {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        // [END get_firestore_instance]

        // [START set_firestore_settings]
        FirebaseFirestoreSettings settings = new FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build();
        db.setFirestoreSettings(settings);

        db.collection("merchants").document(sessionManager.getEmail()).collection("reward_wallet").document("deposit").get()
                .addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                        if(task.isSuccessful()){
                            totalWalletBalance=0L;
                            DocumentSnapshot documentSnapshot = task.getResult();
                            if(documentSnapshot.exists()){
                                totalWalletBalance= (long) documentSnapshot.get("merchant_wallet_amount");
                            }
                        }
                        else{
                            totalWalletBalance=0L;
                        }
                    }
                });


    }

    private void getTotalAmountGifted() {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        // [END get_firestore_instance]

        // [START set_firestore_settings]
        FirebaseFirestoreSettings settings = new FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build();
        db.setFirestoreSettings(settings);

        db.collection("merchants").document(sessionManager.getEmail()).collection("reward_statistics").document("customers").collection("customer_details").get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if(task.isSuccessful()){
                            totalAmountGifted=0L;
                            for (QueryDocumentSnapshot queryDocumentSnapshot:task.getResult()){
                                Long gift_coin_amount=(long) queryDocumentSnapshot.get("gift_coin");
                                totalAmountGifted+=gift_coin_amount;

                            }
                        }
                        else{
                            totalAmountGifted=0L;
                        }
                    }
                });

    }

    private void getNumberOfCustomersGifted() {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        // [END get_firestore_instance]

        // [START set_firestore_settings]
        FirebaseFirestoreSettings settings = new FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build();
        db.setFirestoreSettings(settings);
        db.collection("merchants").document(sessionManager.getEmail()).collection("reward_statistics").document("customers").collection("customer_details").get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if(task.isSuccessful()){
                            numberOfCustomerGifted=0;
                            for (QueryDocumentSnapshot queryDocumentSnapshot:task.getResult()){
                               numberOfCustomerGifted+=1;
                            }
                        }
                        else{
                            numberOfCustomerGifted=0;
                        }
                    }
                });
    }


    public void openFragment(Fragment fragment) {
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.fr_layout_merchant, fragment)
                .addToBackStack(null)
                .commit();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.merchant_menu_main,menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(t.onOptionsItemSelected(item)){
            return true;
        }

        switch (item.getItemId()){

            case R.id.merchant_refresh_page:
                Intent intent = new Intent(this, MerchantActivity.class);
                startActivity(intent);
                return true;
            case R.id.merchant_update_info:
                carouselViewMerchant.setVisibility(View.GONE);
                MerchantInfoUpdate merchantInfoUpdate = new MerchantInfoUpdate();
                openFragment(merchantInfoUpdate);
                return true;

            case R.id.merchant_about_giftin:
                carouselViewMerchant.setVisibility(View.GONE);
                GiftinAboutForMerchant giftinAboutForMerchant = new GiftinAboutForMerchant();
                openFragment(giftinAboutForMerchant);
                return true;

            case R.id.merchant_exit:
                android.app.AlertDialog.Builder builder = new android.app.AlertDialog.Builder(MerchantActivity.this);
                // builder.setTitle("Alert");
                // builder.setIcon(R.drawable.ic_launcher);
                builder.setMessage("   Log out?");
                builder.setCancelable(false);
                builder.setNegativeButton("Cancel",
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int id) {

                            }
                        });

                builder.setNeutralButton("Ok", (dialog, id) -> {
                    mAuth.signOut();
                    sessionManager.clearData();
                    startActivity(new Intent(MerchantActivity.this,SignUpActivity.class));
                    dialog.cancel();

                });
                builder.show();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }


    private void selectDrawerItem(MenuItem menuitem){
        if(menuitem.getItemId()==R.id.navigation_gift_customer_fan){
            carouselViewMerchant.setVisibility(View.GONE);
            GiftACustomerFragment giftACustomer = new GiftACustomerFragment();
            openFragment(giftACustomer);
        }
        else if(menuitem.getItemId() == R.id.navigation_wallet_info){
            carouselViewMerchant.setVisibility(View.GONE);
            WalletInfo walletInfo = new WalletInfo();
            openFragment(walletInfo);
        }
        else if(menuitem.getItemId() == R.id.navigation_merchant_gift_stats) {
            carouselViewMerchant.setVisibility(View.GONE);
            MerchantGiftStatsFragment merchantGiftStatsFragment = new MerchantGiftStatsFragment();
            openFragment(merchantGiftStatsFragment);
        }
        else if(menuitem.getItemId() == R.id.navigation_set_reward_deal) {
            carouselViewMerchant.setVisibility(View.GONE);
            SetRewardDeal setRewardDeal = new SetRewardDeal();
            openFragment(setRewardDeal);
        }
        else if(menuitem.getItemId() == R.id.navigation_view_reward_deal) {
            carouselViewMerchant.setVisibility(View.GONE);
            MerchantStoryList merchantStoryList = new MerchantStoryList();
            openFragment(merchantStoryList);
        }
        else if(menuitem.getItemId() == R.id.navigation_view_reward_deal) {
            carouselViewMerchant.setVisibility(View.GONE);
            MerchantStoryList merchantStoryList = new MerchantStoryList();
            openFragment(merchantStoryList);
        }
        else if(menuitem.getItemId() == R.id.navigation_view_rate_influencer){
            carouselViewMerchant.setVisibility(View.GONE);
            RateInfluencerFragment rateInfluencerFragment = new RateInfluencerFragment();
            openFragment(rateInfluencerFragment);
        }

        else if(menuitem.getItemId() == R.id.navigation_merchant_follow_brands){
            carouselViewMerchant.setVisibility(View.GONE);
            BrandPreferenceFragment brandPreferenceFragment = new BrandPreferenceFragment();
            openFragment(brandPreferenceFragment);
        }
        drawer.close();
    }


    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            try {
                if(sessionManager.getCurrentFragment().equals("CustomerRewardStoriesFragment")){
                    super.onBackPressed();
                }
                else {

                    startActivity(new Intent(MerchantActivity.this, MerchantActivity.class));
                    super.onBackPressed();
                }
            }
            catch (Exception e) {
                //mAuth.signOut();
                //sessionManager.clearData();
                startActivity(new Intent(MerchantActivity.this, SignUpActivity.class));
                finish();
            }

        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        appUpdateManager
                .getAppUpdateInfo()
                .addOnSuccessListener(appUpdateInfo -> {
                    // If the update is downloaded but not installed,
                    // notify the user to complete the update.
                    if (appUpdateInfo.installStatus() == InstallStatus.DOWNLOADED) {
                        popupSnackbarForCompleteUpdate();
                    }
                });
    }


    private void popupSnackbarForCompleteUpdate() {
        Snackbar snackbar =
                Snackbar.make(
                        findViewById(R.id.cl_activity_main_merchant),
                        "An update has just been downloaded.",
                        Snackbar.LENGTH_INDEFINITE);
        snackbar.setAction("RESTART", view -> appUpdateManager.completeUpdate());
        snackbar.setActionTextColor(
                getResources().getColor(R.color.tabColorLight));
        snackbar.show();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if(data!=null){
            Log.d("Data",data.toString());
        }
        if (requestCode == 1) {
            if (resultCode != RESULT_OK) {
                Log.d("UpdateFlowFailed", String.valueOf(resultCode));
                // If the update is cancelled or fails,
                // you can request to start the update again.
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }


    private void getNumberOfFollowers() {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        // [END get_firestore_instance]

        // [START set_firestore_settings]
        FirebaseFirestoreSettings settings = new FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build();
        db.setFirestoreSettings(settings);

        db.collection("merchants").get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            QuerySnapshot result = task.getResult();
                            if (result != null) {
                                List<DocumentSnapshot> eachRes = result.getDocuments();
                                for (int i = 0; i < eachRes.size(); i++) {
                                    counter += 1;
                                    db.collection("merchants").document(eachRes.get(i).getId()).collection("followers").get()
                                            .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                                                @Override
                                                public void onComplete(@NonNull Task<QuerySnapshot> followersTask) {
                                                    if (followersTask.isSuccessful()) {
                                                        QuerySnapshot followersQuerry = followersTask.getResult();
                                                        if (followersQuerry != null) {
                                                            List<DocumentSnapshot> eachFollower = followersQuerry.getDocuments();
                                                            for (int j = 0; j < eachFollower.size(); j++) {
                                                                if (eachFollower.get(j).getId().equals(sessionManager.getEmail())) {
                                                                    following += 1;
                                                                }
                                                            }

                                                            if (counter == result.getDocuments().size()) {
                                                                sessionManager.setFollowingCount(following);

                                                            }
                                                        }
                                                    }

                                                }
                                            });
                                }
                            }
                        }
                    }
                });
    }
}

File: app/src/main/java/com/giftinapp/business/MerchantActivity.kt
package com.giftinapp.business

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.text.Html
import android.util.Log
import android.util.SparseArray
import android.view.*
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.fragment.findNavController
import androidx.navigation.ui.setupWithNavController
import co.paystack.android.PaystackSdk
import com.giftinapp.business.databinding.ActivityMerchantBinding
import com.giftinapp.business.utility.RemoteConfigUtil
import com.giftinapp.business.utility.SessionManager
import com.giftinapp.business.utility.base.BaseActivity
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.tasks.Task
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.synnapps.carouselview.CarouselView
import dagger.hilt.android.AndroidEntryPoint
import java.util.*
import javax.inject.Inject
import kotlin.properties.Delegates

@AndroidEntryPoint
open class MerchantActivity : BaseActivity<ActivityMerchantBinding>() {
    @Inject
    lateinit var firebaseRemoteConfig: FirebaseRemoteConfig

    private lateinit var navController: NavController
    private lateinit var binding: ActivityMerchantBinding
    var bottomNavigation: BottomNavigationView? = null
    var sessionManager: SessionManager? = null
    private var appUpdateManager: AppUpdateManager? = null
    private val mAuth = FirebaseAuth.getInstance()
    var remoteConfigUtil: RemoteConfigUtil? = null
    var btnExploreBrand: Button? = null
    private var imageOne = ""
    private var imageTwo = ""
    private var imageThree = ""
    private var carouselViewMerchant: CarouselView? = null
    private var holderListMerchant = SparseArray<MerchantReportsViewHolder>()
    private var numberOfCustomerGifted by Delegates.notNull<Int>()
    private var totalAmountGifted by Delegates.notNull<Long>()
    var userId = ""
    var totalWalletBalance: Long? = null
    private var t: ActionBarDrawerToggle? = null
    var counter = 0
    var following = 0
    private lateinit var navTextView: TextView


    class MerchantReportsViewHolder {
        lateinit var reportValue: TextView
        lateinit var reportName: TextView
        lateinit var reportIcon: ImageView
    }
    // [END get_firestore_instance]

    // [START set_firestore_settings]
    private val walletBalance: Unit
        get() {
            val db = FirebaseFirestore.getInstance()
            // [END get_firestore_instance]

            // [START set_firestore_settings]
            val settings = FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build()
            db.firestoreSettings = settings
            db.collection("merchants").document(sessionManager!!.getEmail()!!)
                .collection("reward_wallet").document("deposit").get()
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        totalWalletBalance = 0L
                        val documentSnapshot = task.result
                        if (documentSnapshot.exists()) {
                            totalWalletBalance = documentSnapshot["merchant_wallet_amount"] as Long
                        }
                    } else {
                        totalWalletBalance = 0L
                    }
                }
        }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.merchant_menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (t!!.onOptionsItemSelected(item)) {
            return true
        }
        when (item.itemId) {
            R.id.merchant_refresh_page -> {
                val intent = Intent(this, MerchantActivity::class.java)
                startActivity(intent)
                return true
            }

            R.id.merchant_update_info -> {
                navController.navigate(R.id.merchantInfoUpdate)
                return true
            }

            R.id.merchant_about_giftin -> {
                navController.navigate(R.id.giftinAboutForMerchant)
                return true
            }

            R.id.merchant_exit -> {
                showMessageDialog(title = "Log out", message = "Do you want to continue to logout from Brandible?",
                    disMissable = false, hasNegativeBtn = true, negbtnText = "Cancel",posBtnText = "Yes", listener = {
                        mAuth.signOut()
                        sessionManager!!.clearData()
                        startActivity(Intent(this@MerchantActivity, SignUpActivity::class.java))
                        finish()
                    }
                )
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun selectDrawerItem(menuitem: MenuItem) {

        when(menuitem.itemId){
            R.id.navigation_wallet_info->{
                navController.navigate(R.id.walletInfo)
            }
            R.id.navigation_set_reward_deal->{
                navController.navigate(R.id.setRewardDeal)
            }
            R.id.navigation_view_reward_deal->{
                navController.navigate(R.id.merchantStoryList2)
            }
            R.id.navigation_merchant_follow_brands->{
                navController.navigate(R.id.brandPreferenceFragment2)
            }
        }
        binding.merchantNavDrawerLayout.close()
    }

    override fun onBackPressed() {
        if ( binding.merchantNavDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            binding.merchantNavDrawerLayout.closeDrawer(GravityCompat.START)
        } else {
            try {
                if (sessionManager!!.getCurrentFragment() != "CustomerRewardStoriesFragment") {
                    startActivity(Intent(this@MerchantActivity, MerchantActivity::class.java))
                    super.onBackPressed()
                }
            } catch (e: Exception) {
                //mAuth.signOut();
                //sessionManager.clearData();
//                startActivity(new Intent(MerchantActivity.this, SignUpActivity.class));
//                finish();
                //super.onBackPressed()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    override fun onResume() {
        super.onResume()

        if(navController.currentDestination?.id == R.id.giftinAboutForMerchant){
            val intent = Intent(this, MerchantActivity::class.java)
            startActivity(intent)
        }
        if(navController.currentDestination?.id == R.id.merchantInfoUpdate){
            val intent = Intent(this, MerchantActivity::class.java)
            startActivity(intent)
        }
        if(navController.currentDestination?.id == R.id.brandPreferenceFragment2){
            val intent = Intent(this, MerchantActivity::class.java)
            startActivity(intent)
        }
        if(navController.currentDestination?.id == R.id.walletInfo){
            val intent = Intent(this, MerchantActivity::class.java)
            startActivity(intent)
        }
        if(navController.currentDestination?.id == R.id.merchantStoryList2){
            val intent = Intent(this, MerchantActivity::class.java)
            startActivity(intent)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (data != null) {
            Log.d("Data", data.toString())
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun getUserId(){
        val db = FirebaseFirestore.getInstance()
        // [END get_firestore_instance]

        // [START set_firestore_settings]
        val settings = FirebaseFirestoreSettings.Builder()
            .setPersistenceEnabled(true)
            .build()
        db.firestoreSettings = settings

            db.collection("merchants").document(
                    sessionManager!!.getEmail().toString()).get()
                    .addOnCompleteListener { task: Task<DocumentSnapshot> ->
                        if (task.isSuccessful) {
                            val documentSnapshot = task.result
                            if (documentSnapshot.exists()) {

                                userId =
                                    if (documentSnapshot["giftorId"] != null) documentSnapshot["giftorId"]
                                        .toString() else  documentSnapshot.id
                            }
                        }
                    }
        }

    private val numberOfFollowers: Unit
        get() {
            val db = FirebaseFirestore.getInstance()
            // [END get_firestore_instance]

            // [START set_firestore_settings]
            val settings = FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build()
            db.firestoreSettings = settings
            db.collection("merchants").document(sessionManager!!.getEmail()!!)
                .collection("followers").get()
                .addOnCompleteListener { followersTask ->
                    if (followersTask.isSuccessful) {
                        val followersQuerry = followersTask.result
                        if (followersQuerry != null) {
                            following = followersQuerry.size()
                            sessionManager!!.setFollowingCount(following)
                            navTextView.text = resources.getString(
                                R.string.brand_name_and_status,
                               userId,
                                following.toString()
                            )
                        }
                    }
                }
        }


    private fun checkAppUpdate(){
        firebaseRemoteConfig.fetchAndActivate().addOnCompleteListener {
            if (it.isSuccessful) {
                var stringValue = ""

                stringValue = remoteConfigUtil?.getUpdateVersion()?.asString().toString()
                val updateMessage = remoteConfigUtil?.getUpdateMessage()
                val isForced = remoteConfigUtil?.getForceUpdate()?.asBoolean()
                Log.d("Version", stringValue)
                Log.d("Config", BuildConfig.VERSION_CODE.toString())
                if (BuildConfig.VERSION_CODE.toString() != stringValue) {
                    if (isForced == true) {
                        showMessage(false, false, message = updateMessage.toString(),null){
                            openPlayStore()
                        }
                    }else{
                        if (isForced != null) {
                            showMessage(true,true, updateMessage.toString(),"Cancel"){
                                openPlayStore()
                            }
                        }
                    }
                }
            }
        }
    }

    private fun openPlayStore() {
        val appPackageName = packageName

        try {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("market://details?id=$appPackageName")
                )
            )
        } catch (anfe: ActivityNotFoundException) {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("https://play.google.com/store/apps/details?id=$appPackageName")
                )
            )
        }
    }

    companion object {
        private const val MY_REQUEST_CODE = 103
    }

    override fun getActivityBinding(inflater: LayoutInflater): ActivityMerchantBinding {
        binding = ActivityMerchantBinding.inflate(layoutInflater)

        val navHostFrag = supportFragmentManager.findFragmentById(R.id.merchant_act_nav_host_fragment) as NavHostFragment
        navController = navHostFrag.findNavController()

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        binding.merchantNavView.setupWithNavController(navController)

        MobileAds.initialize(this) { initializationStatus ->
            val statusMap =
                initializationStatus.adapterStatusMap
            for (adapterClass in statusMap.keys) {
                val status = statusMap[adapterClass]
                Log.d(
                    "MyApp", String.format(
                        "Adapter name: %s, Description: %s, Latency: %d",
                        adapterClass, status!!.description, status.latency
                    )
                )
            }
        }
        checkAppUpdate()

        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        PaystackSdk.initialize(applicationContext)
        // Returns an intent object that you use to check for an update.
        sessionManager = SessionManager(applicationContext)

        if(sessionManager!!.isFirstTimeLogin()) {
            Handler().postDelayed({
                val i = Intent(this, MerchantActivity::class.java)
                startActivity(i)
                finish()
            }, 3000)
            sessionManager!!.setFirstTimeLogin(false)
        }

        t = ActionBarDrawerToggle(
            this,
            binding.merchantNavDrawerLayout,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_close
        )

        binding.merchantNavDrawerLayout.addDrawerListener(object : DrawerLayout.DrawerListener {
            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {}
            override fun onDrawerOpened(drawerView: View) {
                getUserId()
                numberOfFollowers
//                binding.btnExploreBrand.translationZ = 0f
                Log.d("userId",userId)
            }

            override fun onDrawerClosed(drawerView: View) {
                //binding.btnExploreBrand.translationZ = 2f
                t!!.onDrawerClosed(drawerView)
            }

            override fun onDrawerStateChanged(newState: Int) {

            }
        })

        t!!.syncState()
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding.merchantNavView.setNavigationItemSelectedListener { item: MenuItem ->
            selectDrawerItem(item)
            true
        }
        val headerView = binding.merchantNavView.getHeaderView(0)
        navTextView = headerView.findViewById(R.id.nav_header_textView)
        val navImageView = headerView.findViewById<ImageView>(R.id.nav_header_imageView)
        //        Picasso.get().load(R.drawable.ic_brandible_icon).into(navImageView);
        navTextView.text = resources.getString(
            R.string.brand_name_and_status,
            userId,
            following.toString()
        )
        //btnExploreBrand = findViewById(R.id.btnExploreBrand)
        remoteConfigUtil = RemoteConfigUtil()
        //binding.btnExploreBrand.setOnClickListener(View.OnClickListener { openWebView(remoteConfigUtil!!.getBrandLink()) })

        return binding
    }
}